Neos.Neos/Classes/Domain/Workspace/Workspace.php

```diff
@@ -0,0 +1,442 @@
+use Neos\ContentRepository\Core\Projection\ContentGraph\NodeAggregate;
+use Neos\ContentRepository\Core\Projection\ContentGraph\VisibilityConstraints;
+use Neos\ContentRepository\Core\Projection\Workspace\WorkspaceProjection;
+use Neos\ContentRepository\Core\Projection\Workspace\WorkspaceStatus;
+use Neos\ContentRepository\Core\SharedModel\Exception\NodeAggregateCurrentlyDoesNotExist;
+use Neos\ContentRepository\Core\SharedModel\Exception\WorkspaceDoesNotExist;
+use Neos\ContentRepository\Core\SharedModel\Node\NodeAggregateId;
+use Neos\ContentRepository\Core\SharedModel\Workspace\WorkspaceName;
+use Neos\Flow\Annotations as Flow;
+use Neos\ContentRepository\Core\Feature\WorkspaceRebase\Command\RebaseWorkspace;
+use Neos\ContentRepository\Core\SharedModel\Workspace\ContentStreamId;
+use Neos\Neos\Domain\Service\NodeTypeNameFactory;
+use Neos\Neos\PendingChangesProjection\Change;
+use Neos\Neos\PendingChangesProjection\ChangeFinder;
+
+/**
+ * Neos' workspace model
+ *
+ * @api
+ */
+#[Flow\Proxy(false)]
+final class Workspace
+{
+    public function __construct(
+        public readonly WorkspaceName $name,
+        private ContentStreamId $currentContentStreamId,
+        private WorkspaceStatus $currentStatus,
+        private ?WorkspaceName $currentBaseWorkspaceName,
+        private readonly ContentRepository $contentRepository,
+    ) {
+    }
+
+    public function getCurrentContentStreamId(): ContentStreamId
+    {
+        return $this->currentContentStreamId;
+    }
+
+    public function getCurrentStatus(): WorkspaceStatus
+    {
+        return $this->currentStatus;
+    }
+
+    public function getCurrentBaseWorkspaceName(): ?WorkspaceName
+    {
+        return $this->currentBaseWorkspaceName;
+    }
+
+    public function countAllChanges(): int
+    {
+        $changeFinder = $this->contentRepository->projectionState(ChangeFinder::class);
+        $changes = $changeFinder->findByContentStreamId($this->currentContentStreamId);
+
+        return count($changes);
+    }
+
+    public function countChangesInSite(NodeAggregateId $siteId): int
+    {
+        $ancestorNodeTypeName = NodeTypeNameFactory::forSite();
+        $this->requireNodeToBeOfType(
+            $siteId,
+            $ancestorNodeTypeName
+        );
+
+        $changes = $this->resolveNodeIdsToPublishOrDiscard(
+            $siteId,
+            $ancestorNodeTypeName
+        );
+
+        return count($changes);
+    }
+
+    public function countChangesInDocument(NodeAggregateId $documentId): int
+    {
+        $ancestorNodeTypeName = NodeTypeNameFactory::forDocument();
+        $this->requireNodeToBeOfType(
+            $documentId,
+            $ancestorNodeTypeName
+        );
+
+        $changes = $this->resolveNodeIdsToPublishOrDiscard(
+            $documentId,
+            $ancestorNodeTypeName
+        );
+
+        return count($changes);
+    }
+
+    public function publishAllChanges(): int
```

Comment:
The `: int` return in this interface seems a little out of date with so many value objects flying around, with your approval I could take care of a follow-up. Maybe a `PublishResult` would be in order?

Reply:
Makes sense, I've added the corresponding Publishing and Discarding result objects